tensorflow_probability/python/experimental/mcmc/with_reductions.py

```diff
@@ -0,0 +1,159 @@
+# Copyright 2020 The TensorFlow Probability Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ============================================================================
+"""WithReductions Transition Kernel."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+
+import tensorflow.compat.v2 as tf
+
+from tensorflow_probability.python.mcmc import kernel as kernel_base
+from tensorflow_probability.python.mcmc.internal import util as mcmc_util
+from tensorflow.python.util import nest
+
+
+__all__ = [
+    'WithReductions',
+]
+
+
+class WithReductionsKernelResults(
+    mcmc_util.PrettyNamedTupleMixin,
+    collections.namedtuple('WithReductionsKernelResults',
+                           ['streaming_calculations',
+                            'inner_results'])):
+  """Reducer state and diagnostics for `WithReductions`"""
+  __slots__ = ()
+
+
+class WithReductions(kernel_base.TransitionKernel):
+  """Applies `Reducer`s to stream over MCMC samples.
+  `WithReductions` augments an inner MCMC kernel with
+  side-computations that can read the stream of samples as they
+  are generated.  This is relevant for streaming uses of MCMC,
+  where materializing the entire Markov chain history is undesirable,
+  e.g. due to memory limits.
+
+  One `WithReductions` instance can attach an arbitrary collection
+  of side-computations, each of which must be packaged as a
+  `Reducer`.  `WithReductions` operates by generating a
+  sample with its `inner_kernel`'s `one_step`, then invoking each
+  `Reducer`'s `one_step` method on that sample. The updated reducer
+  states are stored in the `streaming_calculations` field of
+  `WithReductions`' kernel results.
+  """
+
+  def __init__(self, inner_kernel, reducers, name=None):
```

Comment:
I'm surprised to not see a function to transform the outputs of inner_kernel to what's passed to reducers. E.g. if you wanted to compute mean acceptance probability, surely you'd like to be able to do something like:

```python
WithReductions(HMC(), MeanReducer(), lambda state, pkr: pkr.is_accepted)
```

Feel free to defer this to future changes if it seems like a good idea.

Reply:
Seems like a great idea! I'll defer this to the future but I was wondering if you think it would be better to achieve that functionality just with nested reducers in `WithReductions`? For example, make an `acceptance_reducer` whose state is simply `pkr.<some number of inner_results>.is_accepted`, wrap that in an `WithReductions`, then put another `MeanReducer` `WithReductions` around that.

So the onion looks like:
MeanReducer WithReductions
    AcceptanceReducer WithReductions
         inner kernels

Reply:
I think it'd go against the spirit of tfp.mcmc to require creating a brand new subclass for each quantity you wish to take a statistic of. Another idea is to have a compound reducer which takes that lambda.

Reply:
Why not just parameterize the MeanReducer by the function whose mean is to be computed?  That seems like the simplest...